File: RecordingApp/app/src/main/java/wycliffeassociates/recordingapp/Playback/fragments/MarkerToolbarFragment.java
package wycliffeassociates.recordingapp.Playback.fragments;

import android.app.Activity;
import android.app.Fragment;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import wycliffeassociates.recordingapp.Playback.interfaces.MediaController;
import wycliffeassociates.recordingapp.Playback.interfaces.VerseMarkerModeToggler;
import wycliffeassociates.recordingapp.R;
import wycliffeassociates.recordingapp.Utils;
import wycliffeassociates.recordingapp.widgets.PlaybackTimer;

/**
 * Created by sarabiaj on 11/15/2016.
 */

public class MarkerToolbarFragment extends Fragment {

    private ImageView mPlaceMarker;

    public interface OnMarkerPlacedListener {
        void onMarkerPlaced();
    }

    private OnMarkerPlacedListener mOnMarkerPlacedListener;
    private VerseMarkerModeToggler mModeToggleCallback;
    private MediaController mMediaController;
    private ImageButton mPlayBtn;
    private ImageButton mPauseBtn;
    private ImageButton mSkipBackBtn;
    private ImageButton mSkipForwardBtn;
    private TextView mPlaybackElapsed;
    private TextView mPlaybackDuration;
    private PlaybackTimer mTimer;


    public static MarkerToolbarFragment newInstance(){
        MarkerToolbarFragment f = new MarkerToolbarFragment();
        return f;
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        mModeToggleCallback = (VerseMarkerModeToggler) activity;
        mMediaController = (MediaController) activity;
        mOnMarkerPlacedListener = (OnMarkerPlacedListener) activity;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState){
        return inflater.inflate(R.layout.fragment_marker_toolbar, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        findViews();
        initViews();
        initTimer(mPlaybackElapsed, mPlaybackDuration);
    }

    private void findViews(){
        View view = getView();
        mPlayBtn = (ImageButton) view.findViewById(R.id.btn_play);
        mPauseBtn = (ImageButton) view.findViewById(R.id.btn_pause);
        mSkipBackBtn = (ImageButton) view.findViewById(R.id.btn_skip_back);
        mSkipForwardBtn = (ImageButton) view.findViewById(R.id.btn_skip_forward);

        mPlaybackElapsed = (TextView) view.findViewById(R.id.playback_elapsed);
        mPlaybackDuration = (TextView) view.findViewById(R.id.playback_duration);

        mPlaceMarker = (ImageView) view.findViewById(R.id.btn_drop_verse_marker);
    }

    private void initViews(){
        mPlayBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Utils.swapViews(new View[]{mPauseBtn}, new View[]{mPlayBtn});
                mMediaController.onMediaPlay();
            }
        });

        mPauseBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Utils.swapViews(new View[]{mPlayBtn}, new View[]{mPauseBtn});
                mMediaController.onMediaPause();
            }
        });

        mSkipBackBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mMediaController.onSeekBackward();
            }
        });

        mSkipForwardBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mMediaController.onSeekForward();
            }
        });

        mPlaceMarker.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mOnMarkerPlacedListener.onMarkerPlaced();
            }
        });
    }

    private void initTimer(final TextView elapsed, final TextView duration) {
        mTimer = new PlaybackTimer(elapsed, duration);
        mTimer.setElapsed(0);
        mTimer.setDuration(mMediaController.getDuration());
    }

    public void onLocationUpdated(int ms){
        mTimer.setElapsed(ms);
    }

    public void onDurationUpdated(int ms){
        mTimer.setDuration(ms);
    }
}

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/Playback/fragments/MarkerToolbarFragment.kt
package org.wycliffeassociates.translationrecorder.Playback.fragments

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import org.wycliffeassociates.translationrecorder.Playback.interfaces.MediaController
import org.wycliffeassociates.translationrecorder.Playback.interfaces.VerseMarkerModeToggler
import org.wycliffeassociates.translationrecorder.Utils
import org.wycliffeassociates.translationrecorder.databinding.FragmentMarkerToolbarBinding
import org.wycliffeassociates.translationrecorder.widgets.PlaybackTimer

/**
 * Created by sarabiaj on 11/15/2016.
 */
class MarkerToolbarFragment : Fragment() {

    interface OnMarkerPlacedListener {
        fun onMarkerPlaced()
    }

    private var mOnMarkerPlacedListener: OnMarkerPlacedListener? = null
    private var mModeToggleCallback: VerseMarkerModeToggler? = null
    private var mMediaController: MediaController? = null
    private var mTimer: PlaybackTimer? = null

    private var _binding: FragmentMarkerToolbarBinding? = null
    private val binding get() = _binding!!

    override fun onAttach(context: Context) {
        super.onAttach(context)
        mModeToggleCallback = context as VerseMarkerModeToggler
        mMediaController = context as MediaController
        mOnMarkerPlacedListener = context as OnMarkerPlacedListener
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        super.onCreateView(inflater, container, savedInstanceState)
        _binding = FragmentMarkerToolbarBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initViews()
        initTimer(binding.playbackElapsed, binding.playbackDuration)
        if (mMediaController!!.isPlaying) {
            showPauseButton()
        } else {
            showPlayButton()
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    fun showPlayButton() {
        if (_binding == null) return
        Utils.swapViews(
            arrayOf(binding.btnPlay),
            arrayOf(binding.btnPause)
        )
    }

    fun showPauseButton() {
        if (_binding == null) return
        Utils.swapViews(
            arrayOf(binding.btnPause),
            arrayOf(binding.btnPlay)
        )
    }

    private fun initViews() {
        binding.btnPlay.setOnClickListener {
            showPauseButton()
            mMediaController?.onMediaPlay()
        }

        binding.btnPause.setOnClickListener {
            showPlayButton()
            mMediaController?.onMediaPause()
        }

        binding.btnSkipBack.setOnClickListener {
            mMediaController?.onSeekBackward()
        }

        binding.btnSkipForward.setOnClickListener {
            mMediaController?.onSeekForward()
        }

        binding.btnDropVerseMarker.setOnClickListener {
            mOnMarkerPlacedListener?.onMarkerPlaced()
        }
    }

    private fun initTimer(elapsed: TextView?, duration: TextView?) {
        mTimer = PlaybackTimer(elapsed, duration)
        mTimer?.setElapsed(mMediaController!!.locationMs)
        mTimer?.setDuration(mMediaController!!.durationMs)
    }

    fun onLocationUpdated(ms: Int) {
        mTimer?.setElapsed(ms)
    }

    fun onDurationUpdated(ms: Int) {
        mTimer?.setDuration(ms)
    }

    companion object {
        fun newInstance(): MarkerToolbarFragment {
            val f = MarkerToolbarFragment()
            return f
        }
    }
}
